packages/opentelemetry-types/src/distributedcontext/entry.ts

```diff
@@ -0,0 +1,85 @@
+/**
+ * Copyright 2019, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * An Entry is used to label anything that is associated with a specific
+ * operation, such as an HTTP request.
+ */
+export interface Entry {
+  /** The entry key */
+  readonly entryKey: EntryKey;
+  /** The entry value */
+  readonly entryValue: EntryValue;
+  /** The metadata for the entry */
+  readonly entryMetadata: EntryMetadata;
+}
+
+/** EntryKey is the name of the Entry. */
+export interface EntryKey {
+  /**
+   * The name of the key.
+   * Restrictions:
+   *   1. Must contain only printable ASCII (codes between 32 and 126 inclusive)
+   *   2. Must have length greater than zero and less than 256.
+   *   3. Must not be empty.
+   */
+  readonly name: string;
+}
+
+/** A validated entry value. */
+export interface EntryValue {
+  /**
+   * The value of a entry.
+   * Restrictions:
+   *   1. It MUST contain only printable ASCII (codes between 32 and 126)
+   */
+  readonly value: string;
+}
+
+/**
+ * EntryMetadata contains properties associated with an entry. For now only the
+ * property EntryTTL is defined. In future, additional properties may be added
+ * to address specific situations.
+ *
+ * Anytime a sender serializes an entry, sends it over the wire and receiver
+ * deserializes the entry then the entry is considered to have travelled one
+ * hop. There could be one or more proxy(ies) between sender and receiver.
+ * Proxies are treated as transparent entities and they do not create
+ * additional hops.
+ */
+export interface EntryMetadata {
+  /**
+   * For now, only special values of EntryTtl are supported. In future,
+   * additional properties may be added to address specific situations.
+   */
+  readonly entryTtl: EntryTtl;
```

Comment:
What is the use case for this?

Reply:
It sounds like these are completely different things, and the same construct should not be used for them.

* For `0`, simply don't use this feature at all, then it will not be propagated.
* For `-1` this is the equivalent of baggage items in OpenTracing. These are items at the trace level, so they should propagate for the entire trace.
* For `1` the use case you mentioned will always be a single hop. This can be propagated normally using for example a header, which will not be propagated again.
* For `> 0` I don't think there are any use cases.

The proposed API tries to put all of this in the same class, which is very confusing and too verbose. These are different concerns that should be handled differently.